cpp/demo/custom_cell_kernel/main.cpp

```diff
@@ -0,0 +1,138 @@
+// Custom cell kernel (C++)
+//
+// This demo shows how to define a custom cell kernel in C++ and have it
+// assembled into a dolfinx::la::MatrixCSR.
+//
+// .. code-block:: cpp
+
+#include <iostream>
+
+#include <basix/finite-element.h>
+#include <basix/mdspan.hpp>
+#include <basix/quadrature.h>
+#include <cmath>
+#include <dolfinx.h>
+#include <dolfinx/la/MatrixCSR.h>
+#include <dolfinx/la/SparsityPattern.h>
+#include <utility>
+#include <vector>
+
+using namespace dolfinx;
+
+using T = double; // field scalar type
+using U = double; // geometry scalar type
+
+template <typename T, std::size_t ndim>
+using mdspan_t = MDSPAN_IMPL_STANDARD_NAMESPACE::mdspan<
+    T, MDSPAN_IMPL_STANDARD_NAMESPACE::dextents<std::size_t, ndim>>;
+
+// .. code-block:: cpp
+
+int main(int argc, char* argv[])
+{
+  dolfinx::init_logging(argc, argv);
+  MPI_Init(&argc, &argv);
+  {
+    // Create mesh and function space
+    const auto part
+        = mesh::create_cell_partitioner(mesh::GhostMode::shared_facet);
+    auto mesh = std::make_shared<mesh::Mesh<U>>(
+        mesh::create_rectangle<U>(MPI_COMM_WORLD, {{{0.0, 0.0}, {1.0, 1.0}}},
+                                  {10, 10}, mesh::CellType::triangle, part));
+
+    // Create basix element for the field u. This will be used to construct
+    // basis functions inside the custom cell kernel.
+    const int order = 1;
+    const basix::FiniteElement e = basix::create_element<T>(
```

Comment:
No need for `const` - just extra syntactic clutter. (more below that can be removed)